launch_handler-explainer.md

```diff
@@ -0,0 +1,141 @@
+# Web App Launch Handling
+
+## Overview
+
+This document describes a new `launch_handler` manifest member that enables
+web apps to customise their launch behaviour across all types of app launch
+triggers.
+
+
+## Use Cases
+
+- Web apps that are designed to be used in a single window e.g. a music app.
+
+- Web app that capture and handle share target events and user navigations
+  in existing windows without invoking a navigation and losing existing state.
+  E.g. sharing an image to a chat web app could open a contact picker overlayed
+  on top of existing chat content.
+
+
+## Background
+
+There are several ways for a web app window to be opened:
+- [File handling](https://github.com/WICG/file-handling/blob/main/explainer.md)
+- [In scope link capturing](https://github.com/WICG/sw-launch/blob/master/declarative_link_capturing.md)
+- [Note taking](https://wicg.github.io/manifest-incubations/index.html#note_taking-member)
+- Platform specific app launch surface
+- [Protocol handling](https://github.com/MicrosoftEdge/MSEdgeExplainers/blob/main/URLProtocolHandler/explainer.md)
+- [Share target](https://w3c.github.io/web-share-target/)
+- [Shortcuts](https://www.w3.org/TR/appmanifest/#dfn-shortcuts)
+
+Web apps launched via these triggers will open in a new or existing app window
+depending on the user agent platform. There is currently no mechanism for the
+web app to configure this behaviour.
+
+
+## Proposal
+
+- Add a `launch_handler` member to the web app manifest specifying the default
+  client selection and navigation behaviour for web app launches.
+  The shape of this member is as follows:
+  ```
+  "launch_handler": {
+    "route_to": "auto" | "new" | "existing",
+    "navigate_client": "auto" | true | false
+  }
+  ```
+
+  If unspecified then `launch_handler` defaults to
+  `{"route_to": "auto", "navigate_client": "auto"}` whereby the behaviour
+  is up to the user agent.
+
+  Both `route_to` and `navigate_client` also accept a list of values, the
+  first valid value will be used.
+
+  Example manifest that choses to receive all app launches as events in existing
+  web app windows:
+  ```json
+  {
+    "name": "Example app",
+    "start_url": "/index.html",
+    "launch_handler": {
+      "route_to": "existing",
+      "navigate_client": false
+    }
+  }
+  ```
+
+- Enqueue a [`LaunchParams`](
```

Comment:
This is hard to understand if someone hasn't read/used file handling (it implies that `launchQueue` is an existing thing, when actually it's something that file handling introduced, but eventually will be part of *this spec*).

I think you should add a whole section to this explainer for `launchQueue` and `LaunchParams`, copying from file handling if necessary, to explain that while it was first introduced by file handling, it is actually part of the launch handling feature, and will ultimately be specced here, and referenced by file handling and others.

Reply:
Brought in `LaunchParams` and `LaunchQueue` from File Handling.